pkgs/development/beam-modules/rebar3-release.nix

```diff
@@ -27,28 +29,38 @@ let
     (_: v: v != null)
     { inherit setupHook configurePhase buildPhase installPhase; };
 
-  pkg = self: stdenv.mkDerivation (attrs // {
+  rebar3 = rebar3WithPlugins {
+    plugins = buildPlugins;
+  };
+
+  pkg =
+    assert beamDeps != [] -> checkouts == null;
+    self: stdenv.mkDerivation (attrs // {
 
     name = "${name}-${version}";
     inherit version;
 
-    buildInputs = buildInputs ++ [ erlang rebar3 openssl ];
+    buildInputs = buildInputs ++ [ erlang rebar3 openssl ] ++ beamDeps;
     propagatedBuildInputs = [checkouts];
 
     dontStrip = true;
 
     inherit src;
 
-    setupHook = writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib/"
```

Comment:
I removed this since there is nothing in `/lib` anyway

Reply:
One question here. Isn't that how rebar checks for compiled deps?
I'm about to do the same actually for mix release and was thinking about using this exact setup hook to achieve this. Was rebar giving you trouble if you used this ERL_LIBS var ?
Perhaps add a comment down there of why you chose this approach, just to make sure knowledge doesn't get lost.

Reply:
Yea ERL_LIBS (and this hook) is how deps compiled with `buildRebar3` make themselves available to rebar3/erlang. However this derivation is for building an erlang release or escript, which have all deps bundled and aren't intended to be used as dependencies themselves. The names of the derivations don't make this clear, I think what we need is some documentation and some examples of how to use these derivations. I can add some to this PR. Is the nixpkgs repo the place for an example project or should it live somewhere else?

Reply:
I see, what you are saying is essentially that not only dependencies need to be available at build time, they need to be bundled in the release. It's the same for elixir actually.
You don't seem to be copying those dependencies anywhere though?
How does rebar know how to access them? Is that what is happening in that small hook? You are giving the location of the dependencies by overriding the rebar.lock?

Reply:
My understanding of what happens (which might be incomplete) is that `buildRebar3` has [the same setup hook](https://github.com/NixOS/nixpkgs/blob/e097fa08c8c6908d3ed80952e30d6c984241fbd4/pkgs/development/beam-modules/build-rebar3.nix#L45). This ends up in a file in the final derivation under `nix-support/setup-hook`. When this derivation is added to `buildInputs`, as we do here, [the setup hook will be executed](https://github.com/NixOS/nixpkgs/blob/e097fa08c8c6908d3ed80952e30d6c984241fbd4/pkgs/stdenv/generic/setup.sh#L499-L502), adding the libraries to `ERL_LIBS`, where rebar3 is able to find them and copy them into the final escript/release.

Reply:
interesting, I was doubting it would work as easily for mix, but it actually is the exact same behavior.
I made the same PR for mixRelease which is essentially the equivalent of rebar3Relx for mix.
https://github.com/NixOS/nixpkgs/pull/123606

Reply:
I like the idea, however, there is the same `checkouts` concept in mixRelease. The concept of providing dependencies with a fixed output derivation. You need to do the same check, you can't have both FOD and beamDeps. However what if somebody wants an FOD but has buildInputs as well.
Maybe we should disallow FOD entirely? Then I see no more problems with just using buildInputs. It should simplify the logic as well.

Reply:
I think we should move towards removing FOD, but IMHO we should wait until the tooling around the `beamDeps` approach is a bit more mature and well tested. My vote is for keeping `checkouts`/`beamDeps` for now, and having this as a future goal.

Reply:
On second thoughts, perhaps we don't need to be so strict about disallowing checkouts -- we could just document the fact that it won't work?

Reply:
Since changing it will require changes to a few derivations now, I think it should in any case be done in a separate PR (at least for the rebar3 side of things)

Reply:
keeping the FOD was also what came out of the mixRelease conversation.

I personally like the check. It prevent people from running into an error, creating an issue and somebody explaining to them later.

Regarding documentation how about adding a section in the `beam_section.md` to document behavior of `rebarRelx` ? I feel this would be a good way to recommend a way for developpers to use the tool.
Let me know if you would rather have this done in a separate PR (by you or somebody else).
I'm good to merge this as is.

Reply:
Sounds good! Yea the documentation definitely needs some work. If it's OK I'd like to merge this first, then erlang-nix/rebar3_nix#2 (@gleber when you get a minute :)), then perhaps set up an example under `erlang-nix` for each type of rebar3 project people might want to build, so that we can link there from the docs (maybe we can even have the examples in nixpkgs so that CI can check for regressions)?